ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -26,297 +17,401 @@ const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
 ///  - `*_SHIFT`: the number of bits required to left shift a field's value into position (from the
 ///  LSB).
 ///
-/// Bit fiddling for a short operands:
-const SHORT_OPERAND_SIZE: u64 = 18;
-const SHORT_OPERAND_KIND_SIZE: u64 = 3;
-const SHORT_OPERAND_KIND_MASK: u64 = 7;
-const SHORT_OPERAND_VALUE_SIZE: u64 = 15;
-const SHORT_OPERAND_VALUE_SHIFT: u64 = SHORT_OPERAND_KIND_SIZE;
-const SHORT_OPERAND_MAX_VALUE: u64 = !(u64::MAX << SHORT_OPERAND_VALUE_SIZE);
-const SHORT_OPERAND_MASK: u64 = 0x3ffff;
-/// Bit fiddling for instructions.
-const INSTR_ISSHORT_SIZE: u64 = 1;
-const INSTR_ISSHORT_MASK: u64 = 1;
-/// Bit fiddling for short instructions.
-const SHORT_INSTR_OPCODE_MASK: u64 = 0xe;
-const SHORT_INSTR_OPCODE_SHIFT: u64 = INSTR_ISSHORT_SIZE;
-const SHORT_INSTR_FIRST_OPERAND_SHIFT: u64 = INSTR_ISSHORT_SIZE + OPCODE_SIZE;
-
-/// An instruction is identified by its index in the instruction vector.
-#[derive(Copy, Clone)]
-pub(crate) struct InstructionID(usize);
-
-impl InstructionID {
-    pub(crate) fn new(v: usize) -> Self {
-        Self(v)
-    }
+const OPERAND_INDEX_MASK: u16 = 0x7fff;
+
+// The maximum size of an operand we can encode in an instruction is 15 bits.
+const MAX_OPERAND_IDX: u16 = 32768;
+
+// Pre-allocated type indices.
+const TYPE_INDEX_VOID: u16 = 0;
```

Comment:
These three variables might better be an enum with `repr(u16)`?

Reply:
Ah, later on I realise that these are the first 3 elements in a module's type index so an `enum` probably wouldn't be a great idea. A clearer doc comment here would be good.

Reply:
I don't understand this initial comment. The constant is explicitly typed `u16`, so I'm not sure what value annotating a `repr` would add.